### PR TITLE
Tests for DnsSearchDomain functions

### PR DESCRIPTION
**Pull request by jcoglan**
This is James from @neighbourhoodie, I have been working on the STF milestone M1 to improve test coverage, focussing on `resolved`. We have written a lot of tests for this module and we are splitting the work into a number of PRs, one for each test file we've added. Each new test file usually corresponds to a single source file, and exercises the functions in that source file directly.

This PR includes tests for `resolved-dns-search-domain.c`, providing coverage of functions relating to `DnsSearchDomain` objects and lists.

### CODE REVIEW

**Comment by jcoglan** [comment]
I believe I have fixed all the issues with the files in this PR -- asan/ubsan problems, gcc/clang differences, linter checks, etc. and the remaining failing CI tasks are failing for reasons other than the changes here. If that's incorrect and there are other things I still need to change, do let me know what else I need to do.

File: src/resolve/test-dns-search-domain.c [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+#include "resolved-dns-search-domain.h"
+#include "resolved-link.h"
+#include "resolved-manager.h"
+
+#include "log.h"
+#include "tests.h"
+
+static void check_domains(DnsSearchDomain *head, const char **expected, size_t n) {
+        size_t i = 0;
+
```
Comment (bluca) [inline src/resolve/test-dns-search-domain.c:12]: assert(expected)

File: src/resolve/test-dns-search-domain.c [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+        ASSERT_OK(dns_search_domain_new(&manager, &sd, DNS_SEARCH_DOMAIN_SYSTEM, NULL, "local"));
+
+        ASSERT_TRUE(sd->linked);
+        ASSERT_STREQ(DNS_SEARCH_DOMAIN_NAME(sd), "local");
+}
+
+TEST(dns_search_domain_new_system_limit) {
+        Manager manager = {};
+        DnsSearchDomain *sd = NULL;
+        unsigned int i;
```
Comment (bluca) [inline src/resolve/test-dns-search-domain.c]: move declaration inline, and use size_t

File: src/resolve/test-dns-search-domain.c [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+        ASSERT_TRUE(sd->linked);
+        ASSERT_STREQ(DNS_SEARCH_DOMAIN_NAME(sd), "local");
+}
+
+TEST(dns_search_domain_new_link_limit) {
+        Manager manager = {};
+        _cleanup_(link_freep) Link *link = NULL;
+        DnsSearchDomain *sd = NULL;
+        unsigned int i;
```
Comment (bluca) [inline src/resolve/test-dns-search-domain.c]: as above

File: src/resolve/test-dns-search-domain.c [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+        check_domains(manager.search_domains, expected, 2);
+}
+
+TEST(dns_search_domain_unlink_link) {
+        Manager manager = {};
+        Link *link = NULL;
+        _cleanup_(dns_search_domain_unrefp) DnsSearchDomain *sd1 = NULL, *sd3 = NULL;
+        DnsSearchDomain *sd2 = NULL;
+
+        link_new(&manager, &link, 1);
```
Comment (bluca) [inline src/resolve/test-dns-search-domain.c:123]: assert

File: src/resolve/test-dns-search-domain.c [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+        dns_search_domain_new(&manager, &sd1, DNS_SEARCH_DOMAIN_LINK, link, "local");
```
Comment (bluca) [inline src/resolve/test-dns-search-domain.c:126]: also here and other calls below

**Comment by jcoglan** [comment]
@bluca I've made the requested changes and added null checks where needed.

**Review by bluca — APPROVED** [review]

**Comment by bluca** [comment]
merged manually to solve trivial merge conflict

**Pull request closed without merging**